source/CellMeasurer/CellMeasurer.js

```diff
@@ -0,0 +1,222 @@
+     * If this property is not specified the document body will be used.
+     */
+    container: React.PropTypes.oneOfType([
+      React.PropTypes.func,
+      React.PropTypes.node
+    ]),
+
+    /**
+     * Assign a fixed :height in order to measure dynamic text :width only.
+     */
+    height: PropTypes.number,
+
+    /**
+     * Number of rows in grid.
+     */
+    rowCount: PropTypes.number.isRequired,
+
+    /**
+     * Assign a fixed :width in order to measure dynamic text :height only.
+     */
+    width: PropTypes.number
+  };
+
+  constructor (props, state) {
+    super(props, state)
+
+    this._cachedColumnWidths = {}
+    this._cachedRowHeights = {}
+
+    this.getColumnWidth = this.getColumnWidth.bind(this)
+    this.getRowHeight = this.getRowHeight.bind(this)
+    this.resetMeasurements = this.resetMeasurements.bind(this)
+  }
+
+  getColumnWidth ({ index }) {
+    if (this._cachedColumnWidths[index]) {
+      return this._cachedColumnWidths[index]
+    }
+
+    const { rowCount } = this.props
+
+    let maxWidth = 0
+
+    for (var rowIndex = 0; rowIndex < rowCount; rowIndex++) {
+      let { width } = this._measureCell({
+        columnIndex: index,
+        rowIndex
+      })
+
+      maxWidth = Math.max(maxWidth, width)
+    }
+
+    this._cachedColumnWidths[index] = maxWidth
+
+    return maxWidth
+  }
+
+  getRowHeight ({ index }) {
+    if (this._cachedRowHeights[index]) {
+      return this._cachedRowHeights[index]
+    }
+
+    const { rowCount } = this.props
+
+    let maxHeight = 0
+
+    for (var columnIndex = 0; columnIndex < rowCount; columnIndex++) {
+      let { height } = this._measureCell({
+        columnIndex,
+        rowIndex: index
+      })
+
+      maxHeight = Math.max(maxHeight, height)
+    }
+
+    this._cachedRowHeights[index] = maxHeight
+
+    return maxHeight
+  }
+
+  resetMeasurements () {
+    this._cachedColumnWidths = {}
+    this._cachedRowHeights = {}
+  }
+
+  componentDidMount () {
+    this._renderAndMount()
+  }
+
+  componentWillReceiveProps (nextProps) {
+    this._updateDivDimensions(nextProps)
+  }
+
+  componentWillUnmount () {
+    this._unrenderAndDestroy()
+  }
+
+  render () {
+    const { children } = this.props
+
+    return children({
+      getColumnWidth: this.getColumnWidth,
+      getRowHeight: this.getRowHeight,
+      resetMeasurements: this.resetMeasurements
+    })
+  }
+
+  _getContainerNode (props) {
+    const { container } = props
+
+    if (container) {
+      return ReactDOM.findDOMNode(
+        typeof container === 'function'
+          ? container()
+          : container
+      )
+    } else {
+      const node = ReactDOM.findDOMNode(this)
+
+      return node.ownerDocument.body
+    }
+  }
+
+  _measureCell ({
+    columnIndex,
+    rowIndex
+  }) {
+    const { cellRenderer } = this.props
+
+    const rendered = cellRenderer({
+      columnIndex,
+      rowIndex
+    })
+
+    this._div.innerHTML = ReactDOMServer.renderToString(rendered)
```

Comment:
My plan was just to document the limitations of this measurer (eg. it can't be used to measure async-loaded images, fonts that haven't yet been loaded, etc.). I think the async measurer with callback approach you initially used poses a lot more challenges for the library itself though in terms of avoiding unnecessary renders and callbacks.

Let's say we have an RV list that is large enough to display 20 rows at a time. If each one measures async and we have no guarantee what order their callbacks will be invoked in, then we can't know the offset of row 20 until rows 1-19 have completed measuring. Not so awful when we're only talking about 20 rows but if we're talking about hundreds I think it would be very inefficient to have to run through the entire list of sizes and offsets and constantly tweak them as new values came in.

The above problem is way more complex when it comes to `Grid` since we would need to measure every column in a given row to know the rows height (and vice versa to know a given column's width).